File: framework/PythonTools/ParticleInitial.py
class ParticleInitial:
    def __init__(self, event_id, particle_id, particle_type, process, vx, vy, vz, vt, px, py, pz, m, q, eta, phi, pt, p, vertex_primary_id, vertex_secondary_id, generation, sub_particle_id):
        self.event_id = event_id
        self.particle_id = particle_id
        self.particle_type = particle_type
        self.process = process
        self.vx = vx
        self.vy = vy
        self.vz = vz
        self.vt = vt
        self.direction_x = px / p
        self.direction_y = py / p
        self.direction_z = pz / p
        self.px = px
        self.py = py
        self.pz = pz
        self.m = m
        self.q = q
        self.eta = eta
        self.phi = phi
        self.pt = pt
        self.p = p
        self.vertex_primary_id = vertex_primary_id
        self.vertex_secondary_id = vertex_secondary_id
        self.generation = generation
        self.sub_particle_id = sub_particle_id

    def __str__(self):
        return f"ParticleInitial(event_id={self.event_id}, particle_id={self.particle_id}, particle_type={self.particle_type}, process={self.process}, vx={self.vx}, vy={self.vy}, vz={self.vz}, vt={self.vt}, px={self.px}, py={self.py}, pz={self.pz}, m={self.m}, q={self.q}, eta={self.eta}, phi={self.phi}, pt={self.pt}, p={self.p}, vertex_primary_id={self.vertex_primary_id}, vertex_secondary_id={self.vertex_secondary_id}, generation={self.generation}, sub_particle_id={self.sub_particle_id})"

    @staticmethod
    def read_from_csv(path):
        with open(path, 'r') as file:
            lines = file.readlines()

        lines_split = []
        for line in lines:
            lines_split.append([i.strip() for i in line.split(',')])

        particles = []
        for line in lines_split[1:]:
            event_id = int(line[0])
            particle_id = int(line[1])
            particle_type = int(line[2])
            process = int(line[3])
            vx = float(line[4])
            vy = float(line[5])
            vz = float(line[6])
            vt = float(line[7])
            px = float(line[8])
            py = float(line[9])
            pz = float(line[10])
            m = float(line[11])
            q = float(line[12])
            eta = float(line[13])
            phi = float(line[14])
            pt = float(line[15])
            p = float(line[16])
            vertex_primary_id = int(line[17])
            vertex_secondary_id = int(line[18])
            generation = int(line[19])
            sub_particle_id = int(line[20])
            particles.append(ParticleInitial(event_id, particle_id, particle_type, process, vx, vy, vz, vt, px, py, pz, m, q, eta, phi, pt, p, vertex_primary_id, vertex_secondary_id, generation, sub_particle_id))

        return particles
    

File: framework/DataTypes/src/ParticleInitial.cpp
#include "DataTypes/ParticleInitial.h"

DataTypes::ParticleInitial::ParticleInitial(uint32_t eventId, uint64_t particleId, uint32_t particleType, uint32_t process, float vx, float vy, float vz, float vt, float px, float py, float pz, float m, float q, float eta, float phi, float pt, float p, uint32_t vertexPrimaryId, uint32_t vertexSecondaryId, uint32_t particles, uint32_t generation, uint32_t subParticleId)
: eventId_(eventId), particleId_(particleId), particleType_(particleType), process_(process), vx_(vx), vy_(vy), vz_(vz), vt_(vt), directionX_(px / p), directionY_(py / p), directionZ_(pz / p), px_(px), py_(py), pz_(pz), m_(m), q_(q), eta_(eta), phi_(phi), pt_(pt), p_(p), vertexPrimaryId_(vertexPrimaryId), vertexSecondaryId_(vertexSecondaryId), particles_(particles), generation_(generation), subParticleId_(subParticleId) {}

DataTypes::ParticleInitial::ParticleInitial(float vx, float vy, float vz, float px, float py, float pz, float p)
: eventId_(0), particleId_(0), particleType_(0), process_(0), vx_(vx), vy_(vy), vz_(vz), vt_(0), directionX_(px / p), directionY_(py / p), directionZ_(pz / p), px_(px), py_(py), pz_(pz), m_(0), q_(0), eta_(0), phi_(0), pt_(0), p_(p), vertexPrimaryId_(0), vertexSecondaryId_(0), particles_(0), generation_(0), subParticleId_(0) {}
